### ISSUE TITLE
Ethernet: a peer reachable over two interfaces has no way to choose the better path

### ISSUE THREAD

**Issue by Origami74** [jmcorgan/fips#143]
I think this is very much related to our earlier talks about allowing
multiple transports for one peer.

---

*Everything below this line is LLM-written analysis of logs from a live
two-node setup.*

**FIPS version:** `0.6.0-dev` (`rev 6b9faa0c2f`)
**Rust:** `rustc 1.94.1 (e408947bf 2026-03-25)`
**OS:** macOS 14.8.9, arm64

MAC addresses below are placeholders; the topology and counts are from a real
two-node setup.

### What I expected

Two nodes joined by a direct cable use the cable. If they are also both on a
shared wifi LAN, that is a second, worse path — a fallback, not a coin flip.

### What actually happened

The peer's link migrated between the cable and the wifi path on a fixed
cadence, tearing down and re-establishing its session each lap. Because that
peer was also the tree parent, each migration switched parents, invalidated the
downstream coordinate cache and re-announced to every peer.

Nothing was down. Both interfaces stayed present with carrier, `binds: 1`, and
no interface detach event occurred in the whole run — so `show transports`
looked healthy throughout.

### Reproduction

Two nodes, both with an `ethernet` transport on their wifi interface **and** on
a directly-cabled interface, `auto_connect: true` on both (the shipped
default). No further configuration; the flap is the steady state.

### Evidence

Scoped to a single daemon run (transport IDs are reassigned per run, so windows
spanning restarts are not comparable):

```
17 auto-connect dials to one peer in 15 minutes, alternating:
  10  transport:4  02:00:00:00:ca:b1   (cable, en12)
   6  transport:2  02:00:00:00:wi:fi   (wifi,  en0)
   1  transport:4  02:00:00:00:ca:b2
```

```
23:01:07  Peer promoted to active        <peer>
23:02:10  Peer promoted to active        <peer>
23:02:12  Parent switched after first RTT measurement
23:07:06  Parent switched, invalidated downstream coord cache entries,
          announcing to all peers
```

Three parent switches in fifteen minutes. Across hours, promotions appear as
pairs 32-36 s apart. The displaced link reported `etx = 1.0`, `loss = 0.0`.
`show peers` showed the peer's `link_id` climbing (149 → 186) as links were
rebuilt.

---

**Component:** `node.lifecycle` / peer discovery, `peer.active` / link cost.
Contrast with `proto.stp` / parent selection, which already solved the general
problem.

The immediate churn is fixed on `feat/dynamic-interface-binding` (`20132169`).
**Choosing the better path is not.** A node that lands on wifi now stays on
wifi, which is the wrong answer when the other path is a direct cable — a
smaller wrong than a flap, and still a wrong.

---

## The triangle

```
                        ┌───────────────────────┐
                        │  wifi AP / LAN bridge │   ← nobody configured this
                        │  (shared L2 segment)  │      as a FIPS path
                        └─────┬───────────┬─────┘
                              │           │
                     beacons  │           │  beacons
                              │           │
                   en0 ┌──────┴──┐   ┌────┴──────┐ en0
       02:00:00:00:wi:fi│ node A │   │  node B   │
                        └──────┬──┘   └──┬───────┘
                  en12         │         │        peer NIC
                               └─────────┘
                             direct cable
                          02:00:00:00:ca:b1
```

Three vertices, and only two of the edges were intended. The operator plugs in
a cable and believes they have created one path. But both machines were already
on the wifi LAN, and raw-Ethernet beacons flood the L2 broadcast domain — so
the AP silently supplies a second adjacency between exactly the same pair of
nodes.

The result is a peer with two `(transport, MAC)` identities:

| path | transport | interface | peer MAC |
| ---- | --------- | --------- | -------- |
| cable | `transport:4` | `en12` (`dongle3`) | `02:00:00:00:ca:b1` |
| wifi | `transport:2` | `en0` (`wifi`) | `02:00:00:00:wi:fi` |

Both are healthy. Nothing is down, nothing is misconfigured, and no interface
ever detaches — which is why this looks like a working setup right up to the
point someone counts the parent switches.

---

## Why it flaps

Four facts compose into the loop.

**1. Both paths carry beacons, so discovery finds the peer twice.** Each
`(transport, MAC)` pair is a separate candidate from a separate neighbour
buffer. Correct behaviour: they *are* two distinct paths.

**2. FIPS holds one active link per peer.** So the two candidates are rivals,
and exactly one wins.

**3. The gate on re-dialling an active peer only matched the current path.**
`src/node/lifecycle/mod.rs`, before `20132169`:

```rust
// Active peer: skip a candidate whose path is already the
// current, still-fresh one (avoid churning a healthy link).
if self.active_peer_candidate_is_fresh_enough_to_skip(
    &node_addr, std::slice::from_ref(&peer_addr_candidate),
) { continue; }
```

The comment states the intent exactly. The implementation covered only the
same-path case — the one candidate that *cannot* churn anything — and let the
alternate path, the only one that can, straight through. So the non-current
path was dialled once per discovery tick, forever.

**4. A completed dial promotes unconditionally.** There is no cost comparison,
no margin and no hold-down at the path layer, so the newly established link
displaces the incumbent regardless of which is better.

That closes the loop: tick → dial the other path → promote → the other path is
now current → next tick dials the first one back.

```
   ┌──────────────────────────────────────────────┐
   │                                              │
   ▼                                              │
 link held on cable                               │
   │                                              │
   │ discovery tick: wifi candidate is not the    │
   │ current path, so it is not skipped           │
   ▼                                              │
 dial wifi ──► handshake ──► promote ─────────────┤
   │                                              │
   ▼                                              │
 link held on wifi ───────────────────────────────┘
   (and symmetrically back)
```

Session re-established each lap; parent switched each lap when the peer is the
parent.

---

## Why FIPS cannot currently tell which path is better

`ActivePeer::link_cost` (`src/peer/active.rs:745`) is measured-only:

```rust
etx * (1.0 + srtt_ms / 100.0)
```

and it exists per *peer* — for the one link held. The alternate path is
therefore **unmeasured until it is dialled, and dialling it is what displaces
the incumbent.** There is no state in which both paths have a cost, so "switch
only if much better" is unanswerable with today's information.

There is also no transport term in that cost, so nothing distinguishes a cable
from a radio a priori. A 1 Gb/s direct link and a contended wifi hop are
indistinguishable to path selection, and to parent selection until both happen
to have been measured.

---

## Parent selection already solved the general problem

`Stp::evaluate_parent` (`src/proto/stp/state.rs:367`) is the template. It has
every piece path selection lacks:

```rust
// Apply hysteresis: only switch if candidate is significantly better
if best_eff_depth < current_parent_eff * (1.0 - self.parent_hysteresis) {
    return ParentEval::Discretionary(best_peer_id);
}
ParentEval::None
```

- **Cost comparison** — `effective_depth = depth + link_cost`.
- **A multiplicative margin** — `parent_hysteresis`, configured as
  `node.tree.parent_hysteresis`. This is the "much better" requirement.
- **Mandatory vs Discretionary** — a broken path switches unconditionally;
  merely-better is discretionary and veto-gated by `node.tree.hold_down_secs`
  and `node.tree.flap_dampening_secs`.
- **Unmeasured candidates are excluded rather than assumed good:**

```rust
None if peer_costs.is_empty() => 1.0,
None => continue,
```

whose comment reads *"prevents freshly connected peers (no SRTT, default cost
1.0) from appearing artificially cheap. This prevents oscillation from
optimistic defaults."*

That is this bug, one layer up, already diagnosed and fixed. Path selection
never learned it.

---

## Likely fix

Mirror the parent shape for paths. The decision rule transfers directly; the
*input* does not, because parent candidates are all established peers with MMP
data while path candidates are not.

Supplying the missing input, in preference order:

1. **A transport prior.** `EthernetConfig.cost` (default `1.0`, lower is
   better) folded into `link_cost`, so an unmeasured path has a defensible
   starting estimate and the STP-shaped rule can run at all. The cable is
   configured cheap; it wins from wifi on the first evaluation and cannot be
   displaced back, because wifi's prior never beats the margin. A strict
   ordering cannot cycle, which is what keeps the anti-flap property.

   This also fixes parent selection, where a wifi-held and a cable-held peer
   are currently distinguishable only after both are measured.

2. **Hold both links**, let both measure, apply the rule. This is the
   multiple-transports-per-peer idea, arrived at from the other direction —
   this bug is a concrete argument for it rather than a separate problem.

   Architecturally the honest answer: it is the only option in which FIPS ever
   *knows* which path is better, rather than being told or guessing. It is also
   a real change — one-link-per-peer is baked into `addr_to_link`,
   `peers_by_index`, promotion, and the reject accounting — and it subsumes
   option 1, since a prior stops being necessary once both paths are measured.
   A prior would still help choose which to try first, and which to keep when
   the measurements tie.

Either way, keep the liveness gate from `20132169` as the *mandatory* arm: a
dead incumbent is replaced by whatever answers, with no margin and no waiting.

**Not** an option: inferring wired-vs-wireless from the OS. That is
`/sys/class/net/*/wireless` on Linux and considerably worse on Darwin, and it
cannot express "this dongle is my backhaul, that one is not".

---

## Calibration warning

`parent_hysteresis`, `hold_down_secs` and `flap_dampening_secs` exist because
tree parent switching needed all three to stop oscillating. Assume path
selection needs the same and that the values are not transferable — a margin
tuned for `depth + cost` in tree space says nothing about a margin over a
transport prior.

Do not pick these by reasoning. `testing/chaos/scenarios/churn-mixed.yaml`
documents what happens when a threshold is set one step outside a small sample:
it lands inside the real distribution and reddens runs whatever the daemon
does. `ethernet-churn` is the scenario to calibrate against, and it already
carries traffic over Ethernet while interfaces move.

---

## Status

Deferred. This is routing policy, not interface binding, and it wants its own
branch and its own calibration rather than riding along with the presence work
that surfaced it.

What is done: the churn is stopped, and a live link is no longer displaced by
an arbitrary alternate (`20132169`, `active_peer_link_is_live`). Failover is
unaffected — a peer that stops answering goes stale within a heartbeat interval
and every path is dialled again.

What is not: the better path does not win. A node that lands on wifi stays
there while a direct cable sits idle beside it. That is a smaller wrong than a
flap, and still a wrong.

**Operational workaround meanwhile:** stop the peer being reachable twice — set
`auto_connect: false` on the wifi Ethernet transport, or drop the Ethernet
transport on `en0` if the cable is the intended path. Either makes the cable the
only candidate.
